lab-06/README.md

```diff
@@ -0,0 +1,329 @@
+# Design by Introspection
+
+> Inspired by [Bradley Chatha's](https://github.com/BradleyChatha) talk at [DConf 2021](https://www.youtube.com/watch?v=0lo-FOeWecA&list=PLIldXzSkPUXXA0Ephsge-DJEY8o7aZMsR&index=9).
+
+Over the years, a few programming paradigms have been successful enough to enter the casual vocabulary of software engineers: procedural, imperative, object-oriented, functional, generic, declarative.
+But there's a B-list, too, that includes paradigms such as logic, constraint-oriented, and symbolic.
+The point is, there aren't very many of them altogether.
+
+**Design by Introspection (DBI)** is a proposed programming paradigm that is at the same time explosively productive and firmly removed from any of the paradigms considered canon.
+The tenets of Design by Introspection are:
+1. **The rule of optionality:** Component primitives are almost entirely opt-in.
+A given component is required to implement only a modicum of primitives, and all others are optional.
+The component is free to implement any subset of the optional primitives.
+1. **The rule of introspection:** A component user employs introspection on the component to implement its own functionality using the primitives offered by the component.
+1. **The rule of elastic composition:** A component obtained by composing several other components offers capabilities in proportion with the capabilities offered by its individual components.
```

Comment:
These 3 things are very vague and kind of tern. Maybe you should add an example for each. Also, it seems that we talk theoretically about design by introspection, but we don't really back it up with examples and useful patterns. I see that the logger application is interesting, but still, it's not really emphasized on design by introspection. Maybe we can rename this to course to "Metaprogramming"?

Reply:
I agree we should rename it to "Metaprogramming". Using some more advanced patterns would be cool, but I think that would require that this session be held in advance. Food for thought for next year, maybe?